File: app/src/main/java/com/todoroo/astrid/timers/TimerControlSet.java
/**
 * Copyright (c) 2012 Todoroo Inc
 *
 * <p>See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.timers;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.os.SystemClock;
import android.support.annotation.Nullable;
import android.support.v7.app.AlertDialog;
import android.text.TextUtils;
import android.text.format.DateFormat;
import android.text.format.DateUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Chronometer;
import android.widget.ImageView;
import android.widget.TextView;
import butterknife.BindView;
import butterknife.OnClick;
import com.todoroo.andlib.utility.DateUtilities;
import com.todoroo.astrid.data.Task;
import com.todoroo.astrid.ui.TimeDurationControlSet;
import javax.inject.Inject;
import org.tasks.R;
import org.tasks.dialogs.DialogBuilder;
import org.tasks.injection.ForActivity;
import org.tasks.injection.FragmentComponent;
import org.tasks.themes.Theme;
import org.tasks.ui.TaskEditControlFragment;

/**
 * Control Set for managing repeats
 *
 * @author Tim Su <tim@todoroo.com>
 */
public class TimerControlSet extends TaskEditControlFragment {

  public static final int TAG = R.string.TEA_ctrl_timer_pref;
  private static final String EXTRA_STARTED = "extra_started";
  private static final String EXTRA_ESTIMATED = "extra_estimated";
  private static final String EXTRA_ELAPSED = "extra_elapsed";
  @Inject DialogBuilder dialogBuilder;
  @Inject @ForActivity Context context;
  @Inject Theme theme;

  @BindView(R.id.display_row_edit)
  TextView displayEdit;

  @BindView(R.id.timer)
  Chronometer chronometer;

  @BindView(R.id.timer_button)
  ImageView timerButton;

  private TimeDurationControlSet estimated;
  private TimeDurationControlSet elapsed;
  private long timerStarted;
  private AlertDialog dialog;
  private View dialogView;
  private TimerControlSetCallback callback;

  @Nullable
  @Override
  public View onCreateView(
      LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
    View view = super.onCreateView(inflater, container, savedInstanceState);
    int elapsedSeconds;
    int estimatedSeconds;
    if (savedInstanceState == null) {
      timerStarted = task.getTimerStart();
      elapsedSeconds = task.getElapsedSeconds();
      estimatedSeconds = task.getEstimatedSeconds();
    } else {
      timerStarted = savedInstanceState.getLong(EXTRA_STARTED);
      elapsedSeconds = savedInstanceState.getInt(EXTRA_ELAPSED);
      estimatedSeconds = savedInstanceState.getInt(EXTRA_ESTIMATED);
    }

    dialogView = inflater.inflate(R.layout.control_set_timers_dialog, null);
    estimated = new TimeDurationControlSet(context, dialogView, R.id.estimatedDuration, theme);
    elapsed = new TimeDurationControlSet(context, dialogView, R.id.elapsedDuration, theme);
    estimated.setTimeDuration(estimatedSeconds);
    elapsed.setTimeDuration(elapsedSeconds);
    refresh();
    return view;
  }

  @Override
  public void onAttach(Activity activity) {
    super.onAttach(activity);

    callback = (TimerControlSetCallback) activity;
  }

  @Override
  protected void inject(FragmentComponent component) {
    component.inject(this);
  }

  @Override
  public void onSaveInstanceState(Bundle outState) {
    super.onSaveInstanceState(outState);

    outState.putInt(EXTRA_ELAPSED, elapsed.getTimeDurationInSeconds());
    outState.putInt(EXTRA_ESTIMATED, estimated.getTimeDurationInSeconds());
    outState.putLong(EXTRA_STARTED, timerStarted);
  }

  @OnClick(R.id.display_row_edit)
  void openPopup(View view) {
    if (dialog == null) {
      dialog = buildDialog();
    }
    dialog.show();
  }

  private AlertDialog buildDialog() {
    return dialogBuilder
        .newDialog()
        .setView(dialogView)
        .setPositiveButton(android.R.string.ok, (dialog12, which) -> refreshDisplayView())
        .setOnCancelListener(dialog1 -> refreshDisplayView())
        .create();
  }

  @OnClick(R.id.timer_container)
  void timerClicked(View view) {
    if (timerActive()) {
      Task task = callback.stopTimer();
      elapsed.setTimeDuration(task.getElapsedSeconds());
      timerStarted = 0;
      chronometer.stop();
      refreshDisplayView();
    } else {
      Task task = callback.startTimer();
      timerStarted = task.getTimerStart();
      chronometer.start();
    }
    updateChronometer();
  }

  @Override
  protected int getLayout() {
    return R.layout.control_set_timers;
  }

  @Override
  public int getIcon() {
    return R.drawable.ic_timer_24dp;
  }

  @Override
  public int controlId() {
    return TAG;
  }

  @Override
  public boolean hasChanges(Task original) {
    return elapsed.getTimeDurationInSeconds() != original.getElapsedSeconds()
        || estimated.getTimeDurationInSeconds() != original.getEstimatedSeconds();
  }

  @Override
  public void apply(Task task) {
    task.setElapsedSeconds(elapsed.getTimeDurationInSeconds());
    task.setEstimatedSeconds(estimated.getTimeDurationInSeconds());
  }

  private void refresh() {
    refreshDisplayView();
    updateChronometer();
  }

  private void refreshDisplayView() {
    String est = null;
    int estimatedSeconds = estimated.getTimeDurationInSeconds();
    if (estimatedSeconds > 0) {
      est = getString(R.string.TEA_timer_est, DateUtils.formatElapsedTime(estimatedSeconds));
    }
    String elap = null;
    int elapsedSeconds = elapsed.getTimeDurationInSeconds();
    if (elapsedSeconds > 0) {
      elap = getString(R.string.TEA_timer_elap, DateUtils.formatElapsedTime(elapsedSeconds));
    }

    String toDisplay;

    if (!TextUtils.isEmpty(est) && !TextUtils.isEmpty(elap)) {
      toDisplay = est + ", " + elap; // $NON-NLS-1$
    } else if (!TextUtils.isEmpty(est)) {
      toDisplay = est;
    } else if (!TextUtils.isEmpty(elap)) {
      toDisplay = elap;
    } else {
      toDisplay = null;
    }

    displayEdit.setText(toDisplay);
  }

  private void updateChronometer() {
    timerButton.setImageResource(
        timerActive() ? R.drawable.ic_pause_24dp : R.drawable.ic_play_arrow_24dp);

    long elapsed = this.elapsed.getTimeDurationInSeconds() * 1000L;
    if (timerActive()) {
      chronometer.setVisibility(View.VISIBLE);
      elapsed += DateUtilities.now() - timerStarted;
      chronometer.setBase(SystemClock.elapsedRealtime() - elapsed);
      if (elapsed > DateUtilities.ONE_DAY) {
        chronometer.setOnChronometerTickListener(
            cArg -> {
              long t = SystemClock.elapsedRealtime() - cArg.getBase();
              cArg.setText(DateFormat.format("d'd' h:mm", t)); // $NON-NLS-1$
            });
      }
      chronometer.start();
    } else {
      chronometer.setVisibility(View.GONE);
      chronometer.stop();
    }
  }

  private boolean timerActive() {
    return timerStarted > 0;
  }

  public interface TimerControlSetCallback {

    Task stopTimer();

    Task startTimer();
  }
}

File: app/src/main/java/com/todoroo/astrid/timers/TimerControlSet.kt
/*
 * Copyright (c) 2012 Todoroo Inc
 *
 * See the file "LICENSE" for the full license governing this code.
 */
package com.todoroo.astrid.timers

import android.app.Activity
import android.text.format.DateUtils
import android.view.View
import androidx.appcompat.app.AlertDialog
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import androidx.lifecycle.lifecycleScope
import com.todoroo.astrid.ui.TimeDurationControlSet
import dagger.hilt.android.AndroidEntryPoint
import kotlinx.coroutines.launch
import org.tasks.R
import org.tasks.compose.edit.TimerRow
import org.tasks.data.entity.Task
import org.tasks.dialogs.DialogBuilder
import org.tasks.extensions.Context.is24HourFormat
import org.tasks.kmp.org.tasks.time.getTimeString
import org.tasks.time.DateTimeUtils2.currentTimeMillis
import org.tasks.ui.TaskEditControlFragment
import javax.inject.Inject

/**
 * Control Set for managing repeats
 *
 * @author Tim Su <tim></tim>@todoroo.com>
 */
@AndroidEntryPoint
class TimerControlSet : TaskEditControlFragment() {
    @Inject lateinit var activity: Activity
    @Inject lateinit var dialogBuilder: DialogBuilder
    @Inject lateinit var timerPlugin: TimerPlugin
    
    private lateinit var estimated: TimeDurationControlSet
    private lateinit var elapsed: TimeDurationControlSet
    private var dialog: AlertDialog? = null
    private lateinit var dialogView: View

    private fun onRowClick() {
        if (dialog == null) {
            dialog = buildDialog()
        }
        dialog!!.show()
    }

    private fun buildDialog(): AlertDialog {
        return dialogBuilder
                .newDialog()
                .setView(dialogView)
                .setPositiveButton(R.string.ok) { _, _ ->
                    viewModel.estimatedSeconds.value = estimated.timeDurationInSeconds
                    viewModel.elapsedSeconds.value = elapsed.timeDurationInSeconds
                }
                .setOnCancelListener {}
                .create()
    }

    private fun timerClicked() {
        lifecycleScope.launch {
            if (timerActive()) {
                val task = stopTimer()
                viewModel.elapsedSeconds.value = task.elapsedSeconds
                elapsed.setTimeDuration(task.elapsedSeconds)
                viewModel.timerStarted.value = 0
            } else {
                val task = startTimer()
                viewModel.timerStarted.value = task.timerStart
            }
        }
    }

    @Composable
    override fun Content() {
        LaunchedEffect(Unit) {
            dialogView = activity.layoutInflater.inflate(R.layout.control_set_timers_dialog, null)
            estimated = TimeDurationControlSet(activity, dialogView, R.id.estimatedDuration)
            elapsed = TimeDurationControlSet(activity, dialogView, R.id.elapsedDuration)
            estimated.setTimeDuration(viewModel.estimatedSeconds.value)
            elapsed.setTimeDuration(viewModel.elapsedSeconds.value)
        }
        TimerRow(
            started = viewModel.timerStarted.collectAsStateWithLifecycle().value,
            estimated = viewModel.estimatedSeconds.collectAsStateWithLifecycle().value,
            elapsed = viewModel.elapsedSeconds.collectAsStateWithLifecycle().value,
            timerClicked = this@TimerControlSet::timerClicked,
            onClick = this@TimerControlSet::onRowClick,
        )
    }

    private fun timerActive() = viewModel.timerStarted.value > 0

    private suspend fun stopTimer(): Task {
        val model = viewModel.viewState.value.task
        timerPlugin.stopTimer(model)
        val elapsedTime = DateUtils.formatElapsedTime(model.elapsedSeconds.toLong())
        viewModel.addComment(String.format(
            "%s %s\n%s %s",  // $NON-NLS-1$
            getString(R.string.TEA_timer_comment_stopped),
            getTimeString(currentTimeMillis(), requireContext().is24HourFormat),
            getString(R.string.TEA_timer_comment_spent),
            elapsedTime
        ),
            null)
        return model
    }

    private suspend fun startTimer(): Task {
        val model = viewModel.viewState.value.task
        timerPlugin.startTimer(model)
        viewModel.addComment(String.format(
            "%s %s",
            getString(R.string.TEA_timer_comment_started),
            getTimeString(currentTimeMillis(), requireContext().is24HourFormat)
        ),
            null)
        return model
    }

    companion object {
        val TAG = R.string.TEA_ctrl_timer_pref
    }
}
